src/ralph/virtual/admin.py

```diff
@@ -0,0 +1,220 @@
+
+    def get_hypervisor(self, obj):
+        if obj.hypervisor is None:
+            return "Not set"
+        return '<a href="{}">{}</a>'.format(
+            reverse("admin:data_center_datacenterasset_change",
+                    args=(obj.hypervisor.id,)),
+            DataCenterAsset.objects.get(pk=obj.hypervisor).hostname
+        )
+    get_hypervisor.short_description = 'Hypervisor'
+    get_hypervisor.admin_order_field = 'hypervisor__hostname'
+    get_hypervisor.allow_tags = True
+
+    def cloudflavor_name(self, obj):
+        return '<a href="{}">{}</a>'.format(
+            reverse("admin:virtual_cloudflavor_change",
+                    args=(obj.cloudflavor.id,)),
+            obj.cloudflavor.name
+        )
+    cloudflavor_name.short_description = 'Cloud Flavor'
+    cloudflavor_name.admin_order_field = 'cloudflavor__name'
+    cloudflavor_name.allow_tags = True
+
+    def get_ip_addresses(self, obj):
+        return ', '.join(obj.ip_addresses)
+    get_ip_addresses.short_description = 'IP Addresses'
+
+    def get_cloudproject(self, obj):
+        return '<a href="{}">{}</a>'.format(
+            reverse("admin:virtual_cloudproject_change",
+                    args=(obj.parent.id,)),
+            CloudProject.objects.get(pk=obj.parent).name
+        )
+    get_cloudproject.short_description = 'Cloud Project'
+    get_cloudproject.admin_order_field = 'parent'
+    get_cloudproject.allow_tags = True
+
+    def get_cpu(self, obj):
+        return obj.cloudflavor.cores
+    get_cpu.short_description = 'vCPU cores'
+
+    def get_memory(self, obj):
+        return obj.cloudflavor.memory
+    get_memory.short_description = 'RAM size (MiB)'
+
+    def get_disk(self, obj):
+        return obj.cloudflavor.disk/1024
+    get_disk.short_description = 'Disk size (GiB)'
+
+
+@register(CloudFlavor)
+class CloudFlavorAdmin(RalphAdmin):
+    list_display = ['name', 'get_cloudprovider', 'flavor_id']
+    readonly_fields = ['name', 'cloudprovider', 'flavor_id', 'get_cpu',
+                       'get_memory', 'get_disk']
+    fieldsets = (
+        ('Cloud Flavor', {
+            'fields': ['name', 'cloudprovider', 'flavor_id']
+        }),
+        ('Components', {
+            'fields': ['get_cpu', 'get_memory', 'get_disk']
+        }),
+    )
+
+    def has_delete_permission(self, request, obj=None):
+        return False
+
+    def get_cloudprovider(self, obj):
+        return obj.cloudprovider.name
+    get_cloudprovider.short_description = 'Cloud provider'
+    get_cloudprovider.admin_order_field = 'cloudprovider__name'
+
+    def get_cpu(self, obj):
+        return obj.cores
+    get_cpu.short_description = 'vCPU cores'
+
+    def get_memory(self, obj):
+        return obj.memory
+    get_memory.short_description = 'RAM size (MiB)'
+
+    def get_disk(self, obj):
+        return obj.disk/1024
+    get_disk.short_description = 'Disk size (GiB)'
+
+
+@register(CloudProject)
+class CloudProjectAdmin(RalphAdmin):
```

Comment:
check how many sql queries are executed here (the perfect situation is when there is one query for whole table) - probably for each row there are additional queries made (ex. for service_env). try to use `list_select_related`